Supongamos que somos un agente de Airbnb localizado en CDMX, y tenemos que atender peticiones de varios clientes. Tenemos un archivo llamado airbnb_cdmx.csv donde tenemos información de todos los alojamientos de Airbnb en CDMX

In [1]:
import numpy as np
import pandas as pd

``airbnb = pd.read_csv("Direccion donde descargaste el archivo airbnb_cdmx.csv /airbnb_cdmx.csv")
airbnb``

En concreto el dataset tiene las siguientes variables:

* id: el identificador de la propiedad
* host_id: el identificador del dueño de la propiedad
* neighborhood_cleansed: alcaldía en CDMX
* property_type: tipo de propiedad 
* room_type: tipo de habitación
* accommodates: El numero de personas que se pueden alojar en la propiedad price: precio (en pesos mexicanos)
* minimum_nights, maximum_nights: mínimo y máximo de noches
* bedrooms: El número de habitaciones
* beds: número de camas
* cancellation_policy: política de cancelación
* number_of_reviews: El numero de opiniones
* review_scores_accuracy: Puntuacion media del apartamento

In [4]:
airbnb = pd.read_csv("airbnb_cdmx.csv")
airbnb.head()

id  host_id neighbourhood_cleansed property_type        room_type  \
0   35797   153786  Cuajimalpa de Morelos         Villa  Entire home/apt   
1   56074   265650             Cuauhtemoc   Condominium  Entire home/apt   
2   61792   299558             Cuauhtemoc         House     Private room   
3   70644   212109               Coyoacan     Apartment  Entire home/apt   
4  107078   540705         Miguel Hidalgo          Loft  Entire home/apt   

   accommodates  price  minimum_nights  maximum_nights  bedrooms  beds  \
0             2   4500               1               7       1.0   1.0   
1             3    843               4             150       1.0   2.0   
2             2   1359               2              21       1.0   1.0   
3             2   1245               6             180       1.0   1.0   
4             2   4868               4             365       1.0   1.0   

           cancellation_policy  number_of_reviews  review_scores_accuracy  
0                     flexible                  0                     NaN  
1                     moderate                 60                    10.0  
2                     moderate                 52                    10.0  
3                     moderate                102                    10.0  
4  strict_14_with_grace_period                 10                    10.0

## Alicia y su familia

Alicia va a ir a CDMX durante una semana con su marido y sus 2 hijos. Están buscando un apartamento con habitaciones separadas para los padres y los hijos. No les importa donde alojarse o el precio, simplemente quieren tener una experiencia agradable. Esto significa que solo aceptan lugares con más de 10 críticas con una puntuación mayor de 4. Cuando seleccionemos habitaciones para Alicia, tenemos que asegurarnos de ordenar las habitaciones de mejor a peor puntuación. Para aquellas habitaciones que tienen la misma puntuación, debemos mostrar antes aquellas con más críticas. Debemos a lo más la tercera parte del total de posibilidades.

In [ ]:
""" 
noches_estancia = 6
habitaciones >= 2
num_criticas > 10
puntuacion > 4
Ordenar descendente por puntuación
Misma puntuación, ordenar por num_críticas descendente
Mostrar solo 1/3 de los resultados
"""

In [11]:
# Filtrar
# Noches:
# Se quiere quedar exactamente 6 noches.
# Las propiedades que le sirven son aquellas que:
# - no pidan un mínimo mayor a 6 y
# - no tengan un máximo menor a 6.
f1 = airbnb["maximum_nights"] >= 6
f2 = airbnb["minimum_nights"] <= 6
# Habitaciones separadas para papás e hijos
f3 = airbnb["bedrooms"] >= 2
# Mínimo 4 personas
f4 = airbnb["accommodates"] >= 4
# Mínimo 10 críticas
f5 = airbnb["number_of_reviews"] > 10
# Puntuación mínima de 4
f6 = airbnb["review_scores_accuracy"] > 4
# Combinar filtros
opciones = airbnb[f1 & f2 & f3 & f4 & f5 & f6]
opciones

id    host_id neighbourhood_cleansed property_type  \
6        165772     790208         Miguel Hidalgo         House   
9        187030     899360             Cuauhtemoc     Apartment   
12       198778     970609             Cuauhtemoc     Apartment   
15       246057    1288063             Cuauhtemoc     Apartment   
19       256555     790208         Miguel Hidalgo         House   
...         ...        ...                    ...           ...   
19714  42202314  132484396          Benito Juarez     Apartment   
19741  42224748  335393886         Miguel Hidalgo   Condominium   
19771  42249246  332120390          Benito Juarez     Apartment   
19800  42272228  335393886         Miguel Hidalgo   Condominium   
19908  42360548  136203553               Coyoacan     Apartment   

             room_type  accommodates  price  minimum_nights  maximum_nights  \
6      Entire home/apt            14   1245               2             365   
9      Entire home/apt             6   1314               3             365   
12     Entire home/apt             4   1087               2             365   
15     Entire home/apt             5   1699               1              30   
19     Entire home/apt            14   1245               2             365   
...                ...           ...    ...             ...             ...   
19714  Entire home/apt             5    725               2            1125   
19741  Entire home/apt             4   2332               1              30   
19771  Entire home/apt             5   1200               1            1125   
19800  Entire home/apt             4   2255               1              30   
19908  Entire home/apt             4    957               1            1125   

       bedrooms  beds          cancellation_policy  number_of_reviews  \
6           4.0  10.0  strict_14_with_grace_period                266   
9           2.0   3.0                     moderate                 97   
12          2.0   4.0  strict_14_with_grace_period                169   
15          2.0   4.0                     flexible                247   
19          4.0  10.0  strict_14_with_grace_period                264   
...         ...   ...                          ...                ...   
19714       2.0   3.0  strict_14_with_grace_period                 12   
19741       2.0   2.0                     moderate                 21   
19771       2.0   3.0                     flexible                 13   
19800       2.0   2.0                     moderate                 15   
19908       2.0   3.0                     moderate                 14   

       review_scores_accuracy  
6                        10.0  
9                        10.0  
12                        9.0  
15                       10.0  
19                       10.0  
...                       ...  
19714                    10.0  
19741                    10.0  
19771                    10.0  
19800                    10.0  
19908                    10.0  

[3226 rows x 14 columns]

In [17]:
# Ordenar
opciones.sort_values(by=["review_scores_accuracy", "number_of_reviews"],
                     ascending=[False, False],
                     inplace=True)

C:\Users\JorgeZaldivar\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [26]:
# Limitar a la tercera parte
num_opciones = opciones.shape[0] // 3
opciones = opciones.head(num_opciones)
opciones

id    host_id neighbourhood_cleansed property_type  \
2800   14150131   10644799    Venustiano Carranza          Loft   
6024   21052296  151657679              Iztacalco     Apartment   
581     4801896   20440301             Cuauhtemoc     Apartment   
2520   13398036   70937648             Cuauhtemoc     Apartment   
137      997247    5479592             Cuauhtemoc     Apartment   
...         ...        ...                    ...           ...   
10253  29078858   45029087         Miguel Hidalgo     Apartment   
11081  30541361  105748894             Cuauhtemoc     Apartment   
11131  30697022  165046089             Cuauhtemoc     Apartment   
11272  30965631   53361285             Cuauhtemoc     Apartment   
11292  30998957  198471653             Cuauhtemoc     Apartment   

             room_type  accommodates  price  minimum_nights  maximum_nights  \
2800   Entire home/apt             4    473               1            1125   
6024   Entire home/apt             4    540               1            1125   
581    Entire home/apt             4   1993               1             150   
2520   Entire home/apt             4   1200               1            1125   
137    Entire home/apt             4   1812               2            1000   
...                ...           ...    ...             ...             ...   
10253  Entire home/apt             4   1200               2              15   
11081  Entire home/apt             4   1495               2            1125   
11131  Entire home/apt             5   1381               1            1125   
11272  Entire home/apt             7   1902               1            1125   
11292  Entire home/apt             5   1608               2            1125   

       bedrooms  beds          cancellation_policy  number_of_reviews  \
2800        2.0   3.0                     flexible                506   
6024        2.0   2.0                     flexible                424   
581         2.0   2.0                     flexible                365   
2520        2.0   2.0                     moderate                342   
137         2.0   2.0  strict_14_with_grace_period                333   
...         ...   ...                          ...                ...   
10253       2.0   2.0                     moderate                 50   
11081       2.0   2.0                     moderate                 50   
11131       3.0   3.0                     flexible                 50   
11272       3.0   4.0                     moderate                 50   
11292       2.0   3.0  strict_14_with_grace_period                 50   

       review_scores_accuracy  
2800                     10.0  
6024                     10.0  
581                      10.0  
2520                     10.0  
137                      10.0  
...                       ...  
10253                    10.0  
11081                    10.0  
11131                    10.0  
11272                    10.0  
11292                    10.0  

[1075 rows x 14 columns]

## Roberto contra su hermana

Roberto es un casero que tiene una casa en Airbnb. De vez en cuando nos llama preguntando sobre cuales son las críticas de su alojamiento. Hoy está particularmente molesto, ya que su hermana Valentina ha puesto una casa en Airbnb y Roberto quiere asegurarse de que su casa tiene más críticas que las de Valentina. Tenemos que crear un dataframe con las propiedades de ambos. Las id de las casas de Roberto y Valentina son 43856289 y 107078 respectivamente. 

## Diana "la mochilera"

Diana va a CDMX a pasar 3 noches y quiere conocer a gente nueva. Tiene un presupuesto de 1,250 pesos mexicanos para su alojamiento. Debemos buscarle las 10 propiedades más baratas, dandole preferencia a aquellas que sean habitaciones compartidas, y para aquellas viviendas compartidas debemos elegir aquellas con mejor puntuación.